In [512]:
import pandas as pd 

In [513]:
df = pd.read_csv("fifa21 raw data v2.csv", low_memory = False)
df.head(2)

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   

                                           playerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   

   POT                  Club     Contract   Positions Height Weight  \
0   93  \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg   
1   92      \n\n\n\nJuventus  2018 ~ 2022      ST, LW  187cm   83kg   

  Preferred Foot  BOV Best Position        Joined Loan Date End    Value  \
0           Left   93            RW   Jul 1, 2004           NaN  €103.5M   
1          Right   92            ST  Jul 10, 2018           NaN     €63M   

    Wage Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0  €560K        €138.4M        429        85         95                70   
1  €220K         €75.9M        437        84         95                90   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0        347          44             40           93      95         75   
1        353          63             29           95      82         84   

   Composure  Defending  Marking  Standing Tackle  Sliding Tackle  \
0         96         91       32               35              24   
1         95         84       28               32              24   

   Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  \
0           54          6           11          15              14   
1           58          7           11          15              14   

   GK Reflexes  Total Stats  Base Stats  W/F  SM     A/W  D/W   IR  PAC  SHO  \
0            8         2231         466  4 ★  4★  Medium  Low  5 ★   85   92   
1           11         2221         464  4 ★  5★    High  Low  5 ★   89   93   

   PAS  DRI  DEF  PHY Hits  
0   91   95   38   65  771  
1   81   89   35   77  562

In [514]:
df.columns

Index(['ID', 'Name', 'LongName', 'photoUrl', 'playerUrl', 'Nationality', 'Age',
       '↓OVA', 'POT', 'Club', 'Contract', 'Positions', 'Height', 'Weight',
       'Preferred Foot', 'BOV', 'Best Position', 'Joined', 'Loan Date End',
       'Value', 'Wage', 'Release Clause', 'Attacking', 'Crossing', 'Finishing',
       'Heading Accuracy', 'Short Passing', 'Volleys', 'Skill', 'Dribbling',
       'Curve', 'FK Accuracy', 'Long Passing', 'Ball Control', 'Movement',
       'Acceleration', 'Sprint Speed', 'Agility', 'Reactions', 'Balance',
       'Power', 'Shot Power', 'Jumping', 'Stamina', 'Strength', 'Long Shots',
       'Mentality', 'Aggression', 'Interceptions', 'Positioning', 'Vision',
       'Penalties', 'Composure', 'Defending', 'Marking', 'Standing Tackle',
       'Sliding Tackle', 'Goalkeeping', 'GK Diving', 'GK Handling',
       'GK Kicking', 'GK Positioning', 'GK Reflexes', 'Total Stats',
       'Base Stats', 'W/F', 'SM', 'A/W', 'D/W', 'IR', 'PAC', 'SHO', 'PAS',
       'DRI', 'DEF', 

## Drop 'photoUrl' column since playerUrl will contain such information.

In [515]:
df.drop("photoUrl", axis = 1, inplace = True)

In [516]:
df.head(2)

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   

                                           playerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   

   POT                  Club     Contract   Positions Height Weight  \
0   93  \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg   
1   92      \n\n\n\nJuventus  2018 ~ 2022      ST, LW  187cm   83kg   

  Preferred Foot  BOV Best Position        Joined Loan Date End    Value  \
0           Left   93            RW   Jul 1, 2004           NaN  €103.5M   
1          Right   92            ST  Jul 10, 2018           NaN     €63M   

    Wage Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0  €560K        €138.4M        429        85         95                70   
1  €220K         €75.9M        437        84         95                90   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0        347          44             40           93      95         75   
1        353          63             29           95      82         84   

   Composure  Defending  Marking  Standing Tackle  Sliding Tackle  \
0         96         91       32               35              24   
1         95         84       28               32              24   

   Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  \
0           54          6           11          15              14   
1           58          7           11          15              14   

   GK Reflexes  Total Stats  Base Stats  W/F  SM     A/W  D/W   IR  PAC  SHO  \
0            8         2231         466  4 ★  4★  Medium  Low  5 ★   85   92   
1           11         2221         464  4 ★  5★    High  Low  5 ★   89   93   

   PAS  DRI  DEF  PHY Hits  
0   91   95   38   65  771  
1   81   89   35   77  562

In [517]:
df.shape

(18979, 76)

## Capitalize the first letter of the 'playerUrl' column

In [518]:
df.rename(columns={"playerUrl":"PlayerUrl"}, inplace = True)

In [519]:
df.head(2)

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   

                                           PlayerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   

   POT                  Club     Contract   Positions Height Weight  \
0   93  \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg   
1   92      \n\n\n\nJuventus  2018 ~ 2022      ST, LW  187cm   83kg   

  Preferred Foot  BOV Best Position        Joined Loan Date End    Value  \
0           Left   93            RW   Jul 1, 2004           NaN  €103.5M   
1          Right   92            ST  Jul 10, 2018           NaN     €63M   

    Wage Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0  €560K        €138.4M        429        85         95                70   
1  €220K         €75.9M        437        84         95                90   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0        347          44             40           93      95         75   
1        353          63             29           95      82         84   

   Composure  Defending  Marking  Standing Tackle  Sliding Tackle  \
0         96         91       32               35              24   
1         95         84       28               32              24   

   Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  \
0           54          6           11          15              14   
1           58          7           11          15              14   

   GK Reflexes  Total Stats  Base Stats  W/F  SM     A/W  D/W   IR  PAC  SHO  \
0            8         2231         466  4 ★  4★  Medium  Low  5 ★   85   92   
1           11         2221         464  4 ★  5★    High  Low  5 ★   89   93   

   PAS  DRI  DEF  PHY Hits  
0   91   95   38   65  771  
1   81   89   35   77  562

## Change column 'LongName' to 'FullName' and '↓OVA' to OverallRating

In [520]:
df.rename(columns={"LongName":"FullName"}, inplace = True)

In [521]:
df.head(2)

ID               Name                      FullName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   

                                           PlayerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   

   POT                  Club     Contract   Positions Height Weight  \
0   93  \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg   
1   92      \n\n\n\nJuventus  2018 ~ 2022      ST, LW  187cm   83kg   

  Preferred Foot  BOV Best Position        Joined Loan Date End    Value  \
0           Left   93            RW   Jul 1, 2004           NaN  €103.5M   
1          Right   92            ST  Jul 10, 2018           NaN     €63M   

    Wage Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0  €560K        €138.4M        429        85         95                70   
1  €220K         €75.9M        437        84         95                90   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0        347          44             40           93      95         75   
1        353          63             29           95      82         84   

   Composure  Defending  Marking  Standing Tackle  Sliding Tackle  \
0         96         91       32               35              24   
1         95         84       28               32              24   

   Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  \
0           54          6           11          15              14   
1           58          7           11          15              14   

   GK Reflexes  Total Stats  Base Stats  W/F  SM     A/W  D/W   IR  PAC  SHO  \
0            8         2231         466  4 ★  4★  Medium  Low  5 ★   85   92   
1           11         2221         464  4 ★  5★    High  Low  5 ★   89   93   

   PAS  DRI  DEF  PHY Hits  
0   91   95   38   65  771  
1   81   89   35   77  562

In [522]:
df.rename(columns={"↓OVA":"OverallRating"}, inplace = True)

## Expand the columns to access all columns for examination and cleaning

In [524]:
pd.set_option('display.max_columns', None)

In [605]:
df.head(2)

ID               Name                      FullName  \
0  158023.0           L. Messi                  Lionel Messi   
1   20801.0  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   

                                           PlayerUrl Nationality  Age  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35   

   OverallRating   POT          Club     Contract   Positions Height Weight  \
0           93.0  93.0  FC Barcelona  2004 - 2021  RW| ST| CF  170cm   72kg   
1           92.0  92.0      Juventus  2018 - 2022      ST| LW  187cm   83kg   

  Preferred Foot   BOV Best Position     Joined    Value   Wage  \
0           Left  93.0            RW 2004-07-01  €103.5M  €560K   
1          Right  92.0            ST 2018-07-10     €63M  €220K   

  Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0        €138.4M      429.0      85.0       95.0              70.0   
1         €75.9M      437.0      84.0       95.0              90.0   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0           91.0     88.0  470.0       96.0   93.0         94.0          91.0   
1           82.0     86.0  414.0       88.0   81.0         76.0          77.0   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0          96.0     451.0          91.0          80.0     91.0       94.0   
1          92.0     431.0          87.0          91.0     87.0       95.0   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0     95.0  389.0        86.0     68.0     72.0      69.0        94.0   
1     71.0  444.0        94.0     95.0     84.0      78.0        93.0   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0      347.0        44.0           40.0         93.0    95.0       75.0   
1      353.0        63.0           29.0         95.0    82.0       84.0   

   Composure  Defending  Marking  Standing Tackle  Sliding Tackle  \
0       96.0       91.0     32.0             35.0            24.0   
1       95.0       84.0     28.0             32.0            24.0   

   Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  \
0         54.0        6.0         11.0        15.0            14.0   
1         58.0        7.0         11.0        15.0            14.0   

   GK Reflexes  Total Stats  Base Stats W/F SM     A/W  D/W  IR   PAC   SHO  \
0          8.0       2231.0       466.0  4   4  Medium  Low  5   85.0  92.0   
1         11.0       2221.0       464.0  4   5    High  Low  5   89.0  93.0   

    PAS   DRI   DEF   PHY   Hits  
0  91.0  95.0  38.0  65.0  771.0  
1  81.0  89.0  35.0  77.0  562.0

## Remove the '\n\n\n\n' in the club column

In [527]:
df.Club

0               \n\n\n\nFC Barcelona
1                   \n\n\n\nJuventus
2            \n\n\n\nAtlético Madrid
3            \n\n\n\nManchester City
4        \n\n\n\nParis Saint-Germain
                    ...             
18974             \n\n\n\nWuhan Zall
18975        \n\n\n\nOldham Athletic
18976             \n\n\n\nDerry City
18977       \n\n\n\nDalian YiFang FC
18978       \n\n\n\nDalian YiFang FC
Name: Club, Length: 18979, dtype: object

In [528]:
df.Club = df['Club'].str[4:].replace()

In [529]:
df.Club

0               FC Barcelona
1                   Juventus
2            Atlético Madrid
3            Manchester City
4        Paris Saint-Germain
                ...         
18974             Wuhan Zall
18975        Oldham Athletic
18976             Derry City
18977       Dalian YiFang FC
18978       Dalian YiFang FC
Name: Club, Length: 18979, dtype: object

## Examine the dataframe again

In [530]:
df.head(2)

ID               Name                      FullName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   

                                           PlayerUrl Nationality  Age  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35   

   OverallRating  POT          Club     Contract   Positions Height Weight  \
0             93   93  FC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg   
1             92   92      Juventus  2018 ~ 2022      ST, LW  187cm   83kg   

  Preferred Foot  BOV Best Position        Joined Loan Date End    Value  \
0           Left   93            RW   Jul 1, 2004           NaN  €103.5M   
1          Right   92            ST  Jul 10, 2018           NaN     €63M   

    Wage Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0  €560K        €138.4M        429        85         95                70   
1  €220K         €75.9M        437        84         95                90   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0        347          44             40           93      95         75   
1        353          63             29           95      82         84   

   Composure  Defending  Marking  Standing Tackle  Sliding Tackle  \
0         96         91       32               35              24   
1         95         84       28               32              24   

   Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  \
0           54          6           11          15              14   
1           58          7           11          15              14   

   GK Reflexes  Total Stats  Base Stats  W/F  SM     A/W  D/W   IR  PAC  SHO  \
0            8         2231         466  4 ★  4★  Medium  Low  5 ★   85   92   
1           11         2221         464  4 ★  5★    High  Low  5 ★   89   93   

   PAS  DRI  DEF  PHY Hits  
0   91   95   38   65  771  
1   81   89   35   77  562

## With df['Loan Date End'] having 17966 null values out of the 18979 values of the dataframe, it is obvious the column is less informative and important so we drop such unnecessary column to make the data frame neater

In [531]:
df.drop("Loan Date End", axis = 1, inplace = True)

In [532]:
df.head(2)

ID               Name                      FullName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   

                                           PlayerUrl Nationality  Age  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35   

   OverallRating  POT          Club     Contract   Positions Height Weight  \
0             93   93  FC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg   
1             92   92      Juventus  2018 ~ 2022      ST, LW  187cm   83kg   

  Preferred Foot  BOV Best Position        Joined    Value   Wage  \
0           Left   93            RW   Jul 1, 2004  €103.5M  €560K   
1          Right   92            ST  Jul 10, 2018     €63M  €220K   

  Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0        €138.4M        429        85         95                70   
1         €75.9M        437        84         95                90   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0        347          44             40           93      95         75   
1        353          63             29           95      82         84   

   Composure  Defending  Marking  Standing Tackle  Sliding Tackle  \
0         96         91       32               35              24   
1         95         84       28               32              24   

   Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  \
0           54          6           11          15              14   
1           58          7           11          15              14   

   GK Reflexes  Total Stats  Base Stats  W/F  SM     A/W  D/W   IR  PAC  SHO  \
0            8         2231         466  4 ★  4★  Medium  Low  5 ★   85   92   
1           11         2221         464  4 ★  5★    High  Low  5 ★   89   93   

   PAS  DRI  DEF  PHY Hits  
0   91   95   38   65  771  
1   81   89   35   77  562

## Change the '~' in df['Contract'] to '-' which is the standard way to seperate

In [533]:
df['Contract'] = df['Contract'].str.replace("~", "-")

In [534]:
df.head(2)

ID               Name                      FullName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   

                                           PlayerUrl Nationality  Age  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35   

   OverallRating  POT          Club     Contract   Positions Height Weight  \
0             93   93  FC Barcelona  2004 - 2021  RW, ST, CF  170cm   72kg   
1             92   92      Juventus  2018 - 2022      ST, LW  187cm   83kg   

  Preferred Foot  BOV Best Position        Joined    Value   Wage  \
0           Left   93            RW   Jul 1, 2004  €103.5M  €560K   
1          Right   92            ST  Jul 10, 2018     €63M  €220K   

  Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0        €138.4M        429        85         95                70   
1         €75.9M        437        84         95                90   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0        347          44             40           93      95         75   
1        353          63             29           95      82         84   

   Composure  Defending  Marking  Standing Tackle  Sliding Tackle  \
0         96         91       32               35              24   
1         95         84       28               32              24   

   Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  \
0           54          6           11          15              14   
1           58          7           11          15              14   

   GK Reflexes  Total Stats  Base Stats  W/F  SM     A/W  D/W   IR  PAC  SHO  \
0            8         2231         466  4 ★  4★  Medium  Low  5 ★   85   92   
1           11         2221         464  4 ★  5★    High  Low  5 ★   89   93   

   PAS  DRI  DEF  PHY Hits  
0   91   95   38   65  771  
1   81   89   35   77  562

In [535]:
df['Contract'].unique()

array(['2004 - 2021', '2018 - 2022', '2014 - 2023', '2015 - 2023',
       '2017 - 2022', '2017 - 2023', '2018 - 2024', '2014 - 2022',
       '2018 - 2023', '2016 - 2023', '2013 - 2023', '2011 - 2023',
       '2009 - 2022', '2005 - 2021', '2011 - 2021', '2015 - 2022',
       '2017 - 2024', '2010 - 2024', '2012 - 2021', '2019 - 2024',
       '2015 - 2024', '2017 - 2025', '2020 - 2025', '2019 - 2023',
       '2008 - 2023', '2015 - 2021', '2020 - 2022', '2012 - 2022',
       '2016 - 2025', '2013 - 2022', '2011 - 2022', '2012 - 2024',
       '2016 - 2021', '2012 - 2023', '2008 - 2022', '2019 - 2022',
       '2017 - 2021', '2013 - 2024', '2020 - 2024', '2010 - 2022',
       '2020 - 2021', '2011 - 2024', '2020 - 2023', '2014 - 2024',
       '2013 - 2026', '2016 - 2022', '2010 - 2021', '2013 - 2021',
       '2019 - 2025', '2018 - 2025', '2016 - 2024', '2018 - 2021',
       '2009 - 2024', '2007 - 2022', 'Jun 30, 2021 On Loan',
       '2009 - 2021', '2019 - 2021', '2019 - 2026', 'Free', '2012 - 

## Also in df['Positions'], the seperator is a comma, we have to change it to a pipe to make the df clean

In [536]:
df.Positions = df['Positions'].str.replace(",", "|")

In [537]:
df.head(2)

ID               Name                      FullName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   

                                           PlayerUrl Nationality  Age  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35   

   OverallRating  POT          Club     Contract   Positions Height Weight  \
0             93   93  FC Barcelona  2004 - 2021  RW| ST| CF  170cm   72kg   
1             92   92      Juventus  2018 - 2022      ST| LW  187cm   83kg   

  Preferred Foot  BOV Best Position        Joined    Value   Wage  \
0           Left   93            RW   Jul 1, 2004  €103.5M  €560K   
1          Right   92            ST  Jul 10, 2018     €63M  €220K   

  Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0        €138.4M        429        85         95                70   
1         €75.9M        437        84         95                90   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0        347          44             40           93      95         75   
1        353          63             29           95      82         84   

   Composure  Defending  Marking  Standing Tackle  Sliding Tackle  \
0         96         91       32               35              24   
1         95         84       28               32              24   

   Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  \
0           54          6           11          15              14   
1           58          7           11          15              14   

   GK Reflexes  Total Stats  Base Stats  W/F  SM     A/W  D/W   IR  PAC  SHO  \
0            8         2231         466  4 ★  4★  Medium  Low  5 ★   85   92   
1           11         2221         464  4 ★  5★    High  Low  5 ★   89   93   

   PAS  DRI  DEF  PHY Hits  
0   91   95   38   65  771  
1   81   89   35   77  562

## Examine the columns that still have null values

In [538]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 75 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   FullName          18979 non-null  object
 3   PlayerUrl         18979 non-null  object
 4   Nationality       18979 non-null  object
 5   Age               18979 non-null  int64 
 6   OverallRating     18979 non-null  int64 
 7   POT               18979 non-null  int64 
 8   Club              18979 non-null  object
 9   Contract          18979 non-null  object
 10  Positions         18979 non-null  object
 11  Height            18979 non-null  object
 12  Weight            18979 non-null  object
 13  Preferred Foot    18979 non-null  object
 14  BOV               18979 non-null  int64 
 15  Best Position     18979 non-null  object
 16  Joined            18979 non-null  object
 17  Value       

## From the infomation gotten from the .info() above, we can see that the Hits still have about 2595 null values because 16384 - 18979 = 2595.

In [539]:
df.Hits.isnull().sum()

2595

## Get the missen Hits values

## Since the Hits column entails the number of times the players  profile has been viewed on the website, with the column being set to NaN, we have the assume it means their profile has not been viewed on the site, hence we set the value to 0

In [540]:
df.Hits = df.Hits.fillna('0')

In [541]:
df.Hits.isnull().sum()

0

In [542]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 75 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   FullName          18979 non-null  object
 3   PlayerUrl         18979 non-null  object
 4   Nationality       18979 non-null  object
 5   Age               18979 non-null  int64 
 6   OverallRating     18979 non-null  int64 
 7   POT               18979 non-null  int64 
 8   Club              18979 non-null  object
 9   Contract          18979 non-null  object
 10  Positions         18979 non-null  object
 11  Height            18979 non-null  object
 12  Weight            18979 non-null  object
 13  Preferred Foot    18979 non-null  object
 14  BOV               18979 non-null  int64 
 15  Best Position     18979 non-null  object
 16  Joined            18979 non-null  object
 17  Value       

In [543]:
#After changeing the null values in Hits to 0, using .info(), we can see that there are no longer any null values and all we have to do is know if there will be any other futher cleaning to do

## some values in the Hit columns contain 'K', so we change those values to thousand

In [544]:
result = df[df['Hits'].str.contains('K')]


In [545]:
result

ID             Name                       FullName  \
8      231747        K. Mbappé                  Kylian Mbappé   
29     233049        J. Sancho                   Jadon Sancho   
34     212198  Bruno Fernandes  Bruno Miguel Borges Fernandes   
68     231677      M. Rashford                Marcus Rashford   
105    239085       E. Haaland                 Erling Haaland   
106    212462      Alex Telles            Alex Nicolao Telles   
147    229880   A. Wan-Bissaka              Aaron Wan-Bissaka   
416    220651         Angeliño     José Ángel Esmoris Tasende   
544    229558     D. Upamecano                Dayot Upamecano   
687    255475           Antony      Antony Matheus dos Santos   
690    244778          Trincão  Francisco Mota Castro Trincão   
1011   246147     M. Greenwood                Mason Greenwood   
1265   253004        Ansu Fati                 Anssumane Fati   
1380   248243     E. Camavinga              Eduardo Camavinga   
1818   251804          S. Dest                   Sergiño Dest   
2579   231485      A. Tuanzebe                  Axel Tuanzebe   
2757   245371        T. Almada                  Thiago Almada   
3791   251854            Pedri           Pedro González López   
3893   205343    J. Akpa Akpro         Jean-Daniel Akpa Akpro   
4091   254796       N. Madueke                   Noni Madueke   
4096   253283     F. Pellistri              Facundo Pellistri   
4800   251470  C. De Ketelaere           Charles De Ketelaere   
5266   245541         G. Reyna                 Giovanni Reyna   
6159   251852       K. Adeyemi                  Karim Adeyemi   
6235   252371    J. Bellingham                Jude Bellingham   
8116   251570        R. Cherki                   Rayan Cherki   
8401   238071      D. Sterling                 Dujon Sterling   
15819  255611       D. Maldini                 Daniel Maldini   

                                               PlayerUrl    Nationality  Age  \
8      http://sofifa.com/player/231747/kylian-mbappe/...         France   21   
29     http://sofifa.com/player/233049/jadon-sancho/2...        England   20   
34     http://sofifa.com/player/212198/bruno-miguel-b...       Portugal   25   
68     http://sofifa.com/player/231677/marcus-rashfor...        England   22   
105    http://sofifa.com/player/239085/erling-haaland...         Norway   19   
106    http://sofifa.com/player/212462/alex-nicolao-t...         Brazil   27   
147    http://sofifa.com/player/229880/aaron-wan-biss...        England   22   
416    http://sofifa.com/player/220651/jose-angel-esm...          Spain   23   
544    http://sofifa.com/player/229558/dayot-upamecan...         France   21   
687    http://sofifa.com/player/255475/antony-matheus...         Brazil   20   
690    http://sofifa.com/player/244778/francisco-mota...       Portugal   20   
1011   http://sofifa.com/player/246147/mason-greenwoo...        England   18   
1265   http://sofifa.com/player/253004/anssumane-fati...          Spain   17   
1380   http://sofifa.com/player/248243/eduardo-camavi...         France   17   
1818   http://sofifa.com/player/251804/sergino-dest/2...  United States   19   
2579   http://sofifa.com/player/231485/axel-tuanzebe/...        England   22   
2757   http://sofifa.com/player/245371/thiago-almada/...      Argentina   19   
3791   http://sofifa.com/player/251854/pedro-gonzalez...          Spain   17   
3893   http://sofifa.com/player/205343/jean-daniel-ak...    Ivory Coast   27   
4091   http://sofifa.com/player/254796/noni-madueke/2...        England   18   
4096   http://sofifa.com/player/253283/facundo-pellis...        Uruguay   18   
4800   http://sofifa.com/player/251470/charles-de-ket...        Belgium   19   
5266   http://sofifa.com/player/245541/giovanni-reyna...  United States   17   
6159   http://sofifa.com/player/251852/karim-adeyemi/...        Germany   18   
6235   http://sofifa.com/player/252371/jude-bellingha...        England   17   
8116   http://sofifa.com/player/251570/rayan-

In [606]:
# Remove the "." and replace "K" to 00

In [546]:
remove_ch = lambda x: x.replace('.', '').replace('K', '00') if isinstance(x, str) else x


In [547]:
result.Hits = result.Hits.apply(remove_ch)

C:\Users\user\AppData\Local\Temp\ipykernel_1608\1120182670.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  result.Hits = result.Hits.apply(remove_ch)


In [548]:
result.Hits.reset_index()

index  Hits
0       8  1600
1      29  1100
2      34  1500
3      68  3200
4     105  1700
5     106  2300
6     147  1900
7     416   200
8     544  1100
9     687  1300
10    690   300
11   1011  1400
12   1265  2800
13   1380  1800
14   1818  1300
15   2579  8400
16   2757  1100
17   3791  4300
18   3893  1600
19   4091  1200
20   4096  1300
21   4800  2300
22   5266  1800
23   6159  1100
24   6235  1400
25   8116   100
26   8401  2500
27  15819   600

### Some of the thousands came out to be hundred so we change those hundreds to thousands

In [549]:
index_list = [416,690,8116,15819]

In [550]:
rows_to_change = result.loc[index_list]

In [551]:
rows_to_change['Hits'] = rows_to_change['Hits'].apply(lambda x: x.replace('00', '000'))


### After updating the hundreds to thousands, update the output to the result variable which hold values that were formerly 'k's

In [552]:
result.update(rows_to_change)


C:\Users\user\AppData\Local\Temp\ipykernel_1608\1119982295.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  result.update(rows_to_change)
C:\Users\user\AppData\Local\Temp\ipykernel_1608\1119982295.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  result.update(rows_to_change)
C:\Users\user\AppData\Local\Temp\ipykernel_1608\1119982295.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats 

In [553]:
result

ID             Name                       FullName  \
8      231747.0        K. Mbappé                  Kylian Mbappé   
29     233049.0        J. Sancho                   Jadon Sancho   
34     212198.0  Bruno Fernandes  Bruno Miguel Borges Fernandes   
68     231677.0      M. Rashford                Marcus Rashford   
105    239085.0       E. Haaland                 Erling Haaland   
106    212462.0      Alex Telles            Alex Nicolao Telles   
147    229880.0   A. Wan-Bissaka              Aaron Wan-Bissaka   
416    220651.0         Angeliño     José Ángel Esmoris Tasende   
544    229558.0     D. Upamecano                Dayot Upamecano   
687    255475.0           Antony      Antony Matheus dos Santos   
690    244778.0          Trincão  Francisco Mota Castro Trincão   
1011   246147.0     M. Greenwood                Mason Greenwood   
1265   253004.0        Ansu Fati                 Anssumane Fati   
1380   248243.0     E. Camavinga              Eduardo Camavinga   
1818   251804.0          S. Dest                   Sergiño Dest   
2579   231485.0      A. Tuanzebe                  Axel Tuanzebe   
2757   245371.0        T. Almada                  Thiago Almada   
3791   251854.0            Pedri           Pedro González López   
3893   205343.0    J. Akpa Akpro         Jean-Daniel Akpa Akpro   
4091   254796.0       N. Madueke                   Noni Madueke   
4096   253283.0     F. Pellistri              Facundo Pellistri   
4800   251470.0  C. De Ketelaere           Charles De Ketelaere   
5266   245541.0         G. Reyna                 Giovanni Reyna   
6159   251852.0       K. Adeyemi                  Karim Adeyemi   
6235   252371.0    J. Bellingham                Jude Bellingham   
8116   251570.0        R. Cherki                   Rayan Cherki   
8401   238071.0      D. Sterling                 Dujon Sterling   
15819  255611.0       D. Maldini                 Daniel Maldini   

                                               PlayerUrl    Nationality   Age  \
8      http://sofifa.com/player/231747/kylian-mbappe/...         France  21.0   
29     http://sofifa.com/player/233049/jadon-sancho/2...        England  20.0   
34     http://sofifa.com/player/212198/bruno-miguel-b...       Portugal  25.0   
68     http://sofifa.com/player/231677/marcus-rashfor...        England  22.0   
105    http://sofifa.com/player/239085/erling-haaland...         Norway  19.0   
106    http://sofifa.com/player/212462/alex-nicolao-t...         Brazil  27.0   
147    http://sofifa.com/player/229880/aaron-wan-biss...        England  22.0   
416    http://sofifa.com/player/220651/jose-angel-esm...          Spain  23.0   
544    http://sofifa.com/player/229558/dayot-upamecan...         France  21.0   
687    http://sofifa.com/player/255475/antony-matheus...         Brazil  20.0   
690    http://sofifa.com/player/244778/francisco-mota...       Portugal  20.0   
1011   http://sofifa.com/player/246147/mason-greenwoo...        England  18.0   
1265   http://sofifa.com/player/253004/anssumane-fati...          Spain  17.0   
1380   http://sofifa.com/player/248243/eduardo-camavi...         France  17.0   
1818   http://sofifa.com/player/251804/sergino-dest/2...  United States  19.0   
2579   http://sofifa.com/player/231485/axel-tuanzebe/...        England  22.0   
2757   http://sofifa.com/player/245371/thiago-almada/...      Argentina  19.0   
3791   http://sofifa.com/player/251854/pedro-gonzalez...          Spain  17.0   
3893   http://sofifa.com/player/205343/jean-daniel-ak...    Ivory Coast  27.0   
4091   http://sofifa.com/player/254796/noni-madueke/2...        England  18.0   
4096   http://sofifa.com/player/253283/facundo-pellis...        Uruguay  18.0   
4800   http://sofifa.com/player/251470/charles-de-ket...        Belgium  19.0   
5266   http://sofifa.com/player/245541/giovanni-reyna...  United States  17.0   
6159   http://sofifa.com/player/251852/karim-adeyemi/...        Germany  18.0   
6235   http://sofifa.com/player/252371/jude-

In [554]:
df.update(result)


In [555]:
df

ID               Name                      FullName  \
0      158023.0           L. Messi                  Lionel Messi   
1       20801.0  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2      200389.0           J. Oblak                     Jan Oblak   
3      192985.0       K. De Bruyne               Kevin De Bruyne   
4      190871.0          Neymar Jr    Neymar da Silva Santos Jr.   
...         ...                ...                           ...   
18974  247223.0             Xia Ao                        Ao Xia   
18975  258760.0           B. Hough                     Ben Hough   
18976  252757.0        R. McKinley                Ronan McKinley   
18977  243790.0       Wang Zhen'ao                  Zhen'ao Wang   
18978  252520.0          Zhou Xiao                     Xiao Zhou   

                                               PlayerUrl Nationality   Age  \
0      http://sofifa.com/player/158023/lionel-messi/2...   Argentina  33.0   
1      http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal  35.0   
2      http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia  27.0   
3      http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium  29.0   
4      http://sofifa.com/player/190871/neymar-da-silv...      Brazil  28.0   
...                                                  ...         ...   ...   
18974     http://sofifa.com/player/247223/ao-xia/210006/    China PR  21.0   
18975  http://sofifa.com/player/258760/ben-hough/210006/     England  17.0   
18976  http://sofifa.com/player/252757/ronan-mckinley...     England  18.0   
18977  http://sofifa.com/player/243790/zhenao-wang/21...    China PR  20.0   
18978  http://sofifa.com/player/252520/xiao-zhou/210006/    China PR  21.0   

       OverallRating   POT                 Club     Contract   Positions  \
0               93.0  93.0         FC Barcelona  2004 - 2021  RW| ST| CF   
1               92.0  92.0             Juventus  2018 - 2022      ST| LW   
2               91.0  93.0      Atlético Madrid  2014 - 2023          GK   
3               91.0  91.0      Manchester City  2015 - 2023     CAM| CM   
4               91.0  91.0  Paris Saint-Germain  2017 - 2022     LW| CAM   
...              ...   ...                  ...          ...         ...   
18974           47.0  55.0           Wuhan Zall  2018 - 2022          CB   
18975           47.0  67.0      Oldham Athletic  2020 - 2021          CM   
18976           47.0  65.0           Derry City  2019 - 2020          CM   
18977           47.0  57.0     Dalian YiFang FC  2020 - 2022          RW   
18978           47.0  57.0     Dalian YiFang FC  2019 - 2023      CB| LB   

      Height Weight Preferred Foot   BOV Best Position        Joined    Value  \
0      170cm   72kg           Left  93.0            RW   Jul 1, 2004  €103.5M   
1      187cm   83kg          Right  92.0            ST  Jul 10, 2018     €63M   
2      188cm   87kg          Right  91.0            GK  Jul 16, 2014    €120M   
3      181cm   70kg          Right  91.0           CAM  Aug 30, 2015    €129M   
4      175cm   68kg          Right  91.0            LW   Aug 3, 2017    €132M   
...      ...    ...            ...   ...           ...           ...      ...   
18974  178cm   66kg          Right  49.0            CB  Jul 13, 2018    €100K   
18975  175cm   65kg          Right  51.0           CAM   Aug 1, 2020    €130K   
18976  179cm   74kg          Right  49.0           CAM   Mar 8, 2019    €120K   
18977  175cm   69kg          Right  48.0            ST  Sep 22, 2020    €100K   
18978  188cm   75kg           Left  50.0            LB  Jul 29, 2019    €100K   

        Wage Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0      €560K        €138.4M      429.0      85.0       95.0              70.0   
1      €220K         €75.9M      437.0      84.0       95.0              90.0   
2      €125K        €159.4M       95.0      13.0       11.0              15.0   
3      €370K          €161M      407.0      94.0 

In [556]:
df.Hits = df.Hits.astype("int64")

## Some columns have '★' in them so we have to delete that character in the columns to make the data frame more professional

In [557]:
replaced_chr = lambda x: x.replace('★',"")

In [558]:
df.IR = df.IR.apply(replaced_chr)
df.SM = df.SM.apply(replaced_chr)


In [559]:
df['W/F'] = df['W/F'].apply(replaced_chr)

In [560]:
df.head()

ID               Name                      FullName  \
0  158023.0           L. Messi                  Lionel Messi   
1   20801.0  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389.0           J. Oblak                     Jan Oblak   
3  192985.0       K. De Bruyne               Kevin De Bruyne   
4  190871.0          Neymar Jr    Neymar da Silva Santos Jr.   

                                           PlayerUrl Nationality   Age  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina  33.0   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal  35.0   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia  27.0   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium  29.0   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil  28.0   

   OverallRating   POT                 Club     Contract   Positions Height  \
0           93.0  93.0         FC Barcelona  2004 - 2021  RW| ST| CF  170cm   
1           92.0  92.0             Juventus  2018 - 2022      ST| LW  187cm   
2           91.0  93.0      Atlético Madrid  2014 - 2023          GK  188cm   
3           91.0  91.0      Manchester City  2015 - 2023     CAM| CM  181cm   
4           91.0  91.0  Paris Saint-Germain  2017 - 2022     LW| CAM  175cm   

  Weight Preferred Foot   BOV Best Position        Joined    Value   Wage  \
0   72kg           Left  93.0            RW   Jul 1, 2004  €103.5M  €560K   
1   83kg          Right  92.0            ST  Jul 10, 2018     €63M  €220K   
2   87kg          Right  91.0            GK  Jul 16, 2014    €120M  €125K   
3   70kg          Right  91.0           CAM  Aug 30, 2015    €129M  €370K   
4   68kg          Right  91.0            LW   Aug 3, 2017    €132M  €270K   

  Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0        €138.4M      429.0      85.0       95.0              70.0   
1         €75.9M      437.0      84.0       95.0              90.0   
2        €159.4M       95.0      13.0       11.0              15.0   
3          €161M      407.0      94.0       82.0              55.0   
4        €166.5M      408.0      85.0       87.0              62.0   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0           91.0     88.0  470.0       96.0   93.0         94.0          91.0   
1           82.0     86.0  414.0       88.0   81.0         76.0          77.0   
2           43.0     13.0  109.0       12.0   13.0         14.0          40.0   
3           94.0     82.0  441.0       88.0   85.0         83.0          93.0   
4           87.0     87.0  448.0       95.0   88.0         89.0          81.0   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0          96.0     451.0          91.0          80.0     91.0       94.0   
1          92.0     431.0          87.0          91.0     87.0       95.0   
2          30.0     307.0          43.0          60.0     67.0       88.0   
3          92.0     398.0          77.0          76.0     78.0       91.0   
4          95.0     453.0          94.0          89.0     96.0       91.0   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0     95.0  389.0        86.0     68.0     72.0      69.0        94.0   
1     71.0  444.0        94.0     95.0     84.0      78.0        93.0   
2     49.0  268.0        59.0     78.0     41.0      78.0        12.0   
3     76.0  408.0        91.0     63.0     89.0      74.0        91.0   
4     83.0  357.0        80.0     62.0     81.0      50.0        84.0   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0      347.0        44.0           40.0         93.0    95.0       75.0   
1      353.0        63.0           29.0         95.0    82.0       84.0   
2      140.0        34.0           19.0         11.0    65.0       11.0   
3      408.0        76.0           66.0         88.0    94.0       84.0   
4      356.0        51.0           36.0         87.0    

## Change the Joined column to a datetime datatype 

In [561]:
df['Joined'] = pd.to_datetime(df['Joined'])

In [562]:
df['Joined']

0       2004-07-01
1       2018-07-10
2       2014-07-16
3       2015-08-30
4       2017-08-03
           ...    
18974   2018-07-13
18975   2020-08-01
18976   2019-03-08
18977   2020-09-22
18978   2019-07-29
Name: Joined, Length: 18979, dtype: datetime64[ns]

In [563]:
df.head(2)

ID               Name                      FullName  \
0  158023.0           L. Messi                  Lionel Messi   
1   20801.0  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   

                                           PlayerUrl Nationality   Age  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina  33.0   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal  35.0   

   OverallRating   POT          Club     Contract   Positions Height Weight  \
0           93.0  93.0  FC Barcelona  2004 - 2021  RW| ST| CF  170cm   72kg   
1           92.0  92.0      Juventus  2018 - 2022      ST| LW  187cm   83kg   

  Preferred Foot   BOV Best Position     Joined    Value   Wage  \
0           Left  93.0            RW 2004-07-01  €103.5M  €560K   
1          Right  92.0            ST 2018-07-10     €63M  €220K   

  Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0        €138.4M      429.0      85.0       95.0              70.0   
1         €75.9M      437.0      84.0       95.0              90.0   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0           91.0     88.0  470.0       96.0   93.0         94.0          91.0   
1           82.0     86.0  414.0       88.0   81.0         76.0          77.0   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0          96.0     451.0          91.0          80.0     91.0       94.0   
1          92.0     431.0          87.0          91.0     87.0       95.0   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0     95.0  389.0        86.0     68.0     72.0      69.0        94.0   
1     71.0  444.0        94.0     95.0     84.0      78.0        93.0   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0      347.0        44.0           40.0         93.0    95.0       75.0   
1      353.0        63.0           29.0         95.0    82.0       84.0   

   Composure  Defending  Marking  Standing Tackle  Sliding Tackle  \
0       96.0       91.0     32.0             35.0            24.0   
1       95.0       84.0     28.0             32.0            24.0   

   Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  \
0         54.0        6.0         11.0        15.0            14.0   
1         58.0        7.0         11.0        15.0            14.0   

   GK Reflexes  Total Stats  Base Stats W/F SM     A/W  D/W  IR   PAC   SHO  \
0          8.0       2231.0       466.0  4   4  Medium  Low  5   85.0  92.0   
1         11.0       2221.0       464.0  4   5    High  Low  5   89.0  93.0   

    PAS   DRI   DEF   PHY  Hits  
0  91.0  95.0  38.0  65.0   771  
1  81.0  89.0  35.0  77.0   562

## The age column is in a float data type, change it to an integer instead because age cannot be in a float data type

In [603]:
df.Age = df['Age'].astype("int32")

## Work on the height and weight column

# Height Cleaning

In [565]:
df.Height.unique()

array(['170cm', '187cm', '188cm', '181cm', '175cm', '184cm', '191cm',
       '178cm', '193cm', '185cm', '199cm', '173cm', '168cm', '176cm',
       '177cm', '183cm', '180cm', '189cm', '179cm', '195cm', '172cm',
       '182cm', '186cm', '192cm', '165cm', '194cm', '167cm', '196cm',
       '163cm', '190cm', '174cm', '169cm', '171cm', '197cm', '200cm',
       '166cm', '6\'2"', '164cm', '198cm', '6\'3"', '6\'5"', '5\'11"',
       '6\'4"', '6\'1"', '6\'0"', '5\'10"', '5\'9"', '5\'6"', '5\'7"',
       '5\'4"', '201cm', '158cm', '162cm', '161cm', '160cm', '203cm',
       '157cm', '156cm', '202cm', '159cm', '206cm', '155cm'], dtype=object)

In [566]:
#There are some feet in the height column that is supposed to be in centimeter, we have to get those values that are in feet
feet = df[df['Height'].str.contains('"')]

In [567]:
feet

ID              Name                          FullName  \
793  110677.0              Moyá           Miguel Ángel Moyá Rumbo   
847  166706.0         M. Škrtel                     Martin Škrtel   
848  189068.0           B. Dost                          Bas Dost   
860  198031.0   Sérgio Oliveira  Sérgio Miguel Relvas de Oliveira   
861  190243.0           M. Hitz                       Marwin Hitz   
862  194222.0         Laguardia         Víctor Laguardia Cisneros   
863  194209.0       Y. El Arabi                  Youssef El Arabi   
864  193601.0       G. Pezzella                   Germán Pezzella   
865  193352.0      R. Rodríguez                 Ricardo Rodríguez   
866  193152.0          Herrerín              Iago Herrerín Buisán   
867  192546.0        E. Stevens                      Enda Stevens   
868  192123.0           C. Wood                        Chris Wood   
869  192086.0        A. Kokorin                  Alexandr Kokorin   
871  192045.0      L. Rodríguez                    Luis Rodríguez   
872  189280.0       A. Westwood                   Ashley Westwood   
873  196978.0         C. Wilson                     Callum Wilson   
874  188988.0        M. Lanzini                    Manuel Lanzini   
875  173909.0        K. Boateng              Kevin-Prince Boateng   
876  188427.0          S. Lulić                       Senad Lulić   
877  140293.0        A. Mirante                   Antonio Mirante   
878  188268.0         L. Acosta                    Lautaro Acosta   
879  188042.0         M. Suárez                     Matías Suárez   
880  188041.0    Rafael Carioca           Rafael de Souza Pereira   
881  186547.0      M. Musacchio                   Mateo Musacchio   
883  183353.0       Víctor Díaz          Víctor David Díaz Miguel   
884  178250.0           P. Diop                    Papakouli Diop   
885  178213.0         E. Capoue                    Etienne Capoue   
886  175254.0        F. Caicedo                    Felipe Caicedo   
887  183899.0         P. Piatti                      Pablo Piatti   
888  173221.0       A. Candreva                  Antonio Candreva   
889  158626.0        M. Debuchy                   Mathieu Debuchy   
890  143001.0          C. Tévez                      Carlos Tévez   
891  173208.0       D. Criscito                 Domenico Criscito   
892  143121.0           J. Sosa                         José Sosa   
893  154472.0      Jorge Molina                Jorge Molina Vidal   
894  146296.0  Andrés Fernández   Andrés Eduardo Fernández Moreno   
895  171791.0        José Fonte        José Miguel da Rocha Fonte   
896  171897.0       A. Guardado                   Andrés Guardado   
897  172962.0          V. Moses                      Victor Moses   
898  142707.0          L. López                    Lisandro López   

                                             PlayerUrl          Nationality  \
793  http://sofifa.com/player/110677/miguel-angel-m...                Spain   
847  http://sofifa.com/player/166706/martin-skrtel/...             Slovakia   
848   http://sofifa.com/player/189068/bas-dost/210005/          Netherlands   
860  http://sofifa.com/player/198031/sergio-miguel-...             Portugal   
861  http://sofifa.com/player/190243/marwin-hitz/21...          Switzerland   
862  http://sofifa.com/player/194222/victor-laguard...                Spain   
863  http://sofifa.com/player/194209/youssef-el-ara...              Morocco   
864  http://sofifa.com/player/193601/german-pezzell...            Argentina   
865  http://sofifa.com/player/193352/ricardo-rodrig...          Switzerland   
866  http://sofifa.com/player/193152/iago-herrerin-...                Spain   
867  http://sofifa.com/player/192546/enda-stevens/2...  Republic of Ireland   
868  http://sofifa.com/player/192123/chris-wood/210...          New Zealand   
869  http://sofifa.com/player/192086/alexandr-kokor...               Russia   
871  http://sofifa.com/player/192045/luis-rodriguez...               Mexico

In [569]:
# Remove the characters that makes the values a feet 
replace = lambda x: x.replace("'",".").replace('"',' ')

In [570]:
feet['Height'] = feet['Height'].apply(replace).astype(float)

C:\Users\user\AppData\Local\Temp\ipykernel_1608\2730111519.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  feet['Height'] = feet['Height'].apply(replace).astype(float)


In [571]:
#Using the lambda function, change the feets to centimeter
rp = lambda x: x * 30.48
feet['Height'] = feet['Height'].apply(rp).astype('int64')

C:\Users\user\AppData\Local\Temp\ipykernel_1608\1425456056.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  feet['Height'] = feet['Height'].apply(rp).astype('int64')


In [572]:
feet

ID              Name                          FullName  \
793  110677.0              Moyá           Miguel Ángel Moyá Rumbo   
847  166706.0         M. Škrtel                     Martin Škrtel   
848  189068.0           B. Dost                          Bas Dost   
860  198031.0   Sérgio Oliveira  Sérgio Miguel Relvas de Oliveira   
861  190243.0           M. Hitz                       Marwin Hitz   
862  194222.0         Laguardia         Víctor Laguardia Cisneros   
863  194209.0       Y. El Arabi                  Youssef El Arabi   
864  193601.0       G. Pezzella                   Germán Pezzella   
865  193352.0      R. Rodríguez                 Ricardo Rodríguez   
866  193152.0          Herrerín              Iago Herrerín Buisán   
867  192546.0        E. Stevens                      Enda Stevens   
868  192123.0           C. Wood                        Chris Wood   
869  192086.0        A. Kokorin                  Alexandr Kokorin   
871  192045.0      L. Rodríguez                    Luis Rodríguez   
872  189280.0       A. Westwood                   Ashley Westwood   
873  196978.0         C. Wilson                     Callum Wilson   
874  188988.0        M. Lanzini                    Manuel Lanzini   
875  173909.0        K. Boateng              Kevin-Prince Boateng   
876  188427.0          S. Lulić                       Senad Lulić   
877  140293.0        A. Mirante                   Antonio Mirante   
878  188268.0         L. Acosta                    Lautaro Acosta   
879  188042.0         M. Suárez                     Matías Suárez   
880  188041.0    Rafael Carioca           Rafael de Souza Pereira   
881  186547.0      M. Musacchio                   Mateo Musacchio   
883  183353.0       Víctor Díaz          Víctor David Díaz Miguel   
884  178250.0           P. Diop                    Papakouli Diop   
885  178213.0         E. Capoue                    Etienne Capoue   
886  175254.0        F. Caicedo                    Felipe Caicedo   
887  183899.0         P. Piatti                      Pablo Piatti   
888  173221.0       A. Candreva                  Antonio Candreva   
889  158626.0        M. Debuchy                   Mathieu Debuchy   
890  143001.0          C. Tévez                      Carlos Tévez   
891  173208.0       D. Criscito                 Domenico Criscito   
892  143121.0           J. Sosa                         José Sosa   
893  154472.0      Jorge Molina                Jorge Molina Vidal   
894  146296.0  Andrés Fernández   Andrés Eduardo Fernández Moreno   
895  171791.0        José Fonte        José Miguel da Rocha Fonte   
896  171897.0       A. Guardado                   Andrés Guardado   
897  172962.0          V. Moses                      Victor Moses   
898  142707.0          L. López                    Lisandro López   

                                             PlayerUrl          Nationality  \
793  http://sofifa.com/player/110677/miguel-angel-m...                Spain   
847  http://sofifa.com/player/166706/martin-skrtel/...             Slovakia   
848   http://sofifa.com/player/189068/bas-dost/210005/          Netherlands   
860  http://sofifa.com/player/198031/sergio-miguel-...             Portugal   
861  http://sofifa.com/player/190243/marwin-hitz/21...          Switzerland   
862  http://sofifa.com/player/194222/victor-laguard...                Spain   
863  http://sofifa.com/player/194209/youssef-el-ara...              Morocco   
864  http://sofifa.com/player/193601/german-pezzell...            Argentina   
865  http://sofifa.com/player/193352/ricardo-rodrig...          Switzerland   
866  http://sofifa.com/player/193152/iago-herrerin-...                Spain   
867  http://sofifa.com/player/192546/enda-stevens/2...  Republic of Ireland   
868  http://sofifa.com/player/192123/chris-wood/210...          New Zealand   
869  http://sofifa.com/player/192086/alexandr-kokor...               Russia   
871  http://sofifa.com/player/192045/luis-rodriguez...               Mexico

In [573]:
feet['Height'] = feet['Height'].astype('str')

C:\Users\user\AppData\Local\Temp\ipykernel_1608\137004716.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  feet['Height'] = feet['Height'].astype('str')


In [578]:
feet['Height']

793    188
847    192
848    198
860    155
861    195
862    185
863    182
864    185
865    155
866    188
867    182
868    192
869    182
871    155
872    179
873    155
874    170
875    185
876    182
877    195
878    173
879    182
880    155
881    155
883    182
884    155
885    192
886    182
887    164
888    155
889    179
890    173
891    182
892    155
893    188
894    188
895    192
896    173
897    155
898    179
Name: Height, dtype: object

In [579]:
# Add the suffix 'cm' to make the values match those in the main dataframe

In [580]:
func = lambda x: x + "cm"

In [581]:
feet.Height = feet["Height"].apply(func)

C:\Users\user\AppData\Local\Temp\ipykernel_1608\1250980559.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  feet.Height = feet["Height"].apply(func)


In [582]:
feet

ID              Name                          FullName  \
793  110677.0              Moyá           Miguel Ángel Moyá Rumbo   
847  166706.0         M. Škrtel                     Martin Škrtel   
848  189068.0           B. Dost                          Bas Dost   
860  198031.0   Sérgio Oliveira  Sérgio Miguel Relvas de Oliveira   
861  190243.0           M. Hitz                       Marwin Hitz   
862  194222.0         Laguardia         Víctor Laguardia Cisneros   
863  194209.0       Y. El Arabi                  Youssef El Arabi   
864  193601.0       G. Pezzella                   Germán Pezzella   
865  193352.0      R. Rodríguez                 Ricardo Rodríguez   
866  193152.0          Herrerín              Iago Herrerín Buisán   
867  192546.0        E. Stevens                      Enda Stevens   
868  192123.0           C. Wood                        Chris Wood   
869  192086.0        A. Kokorin                  Alexandr Kokorin   
871  192045.0      L. Rodríguez                    Luis Rodríguez   
872  189280.0       A. Westwood                   Ashley Westwood   
873  196978.0         C. Wilson                     Callum Wilson   
874  188988.0        M. Lanzini                    Manuel Lanzini   
875  173909.0        K. Boateng              Kevin-Prince Boateng   
876  188427.0          S. Lulić                       Senad Lulić   
877  140293.0        A. Mirante                   Antonio Mirante   
878  188268.0         L. Acosta                    Lautaro Acosta   
879  188042.0         M. Suárez                     Matías Suárez   
880  188041.0    Rafael Carioca           Rafael de Souza Pereira   
881  186547.0      M. Musacchio                   Mateo Musacchio   
883  183353.0       Víctor Díaz          Víctor David Díaz Miguel   
884  178250.0           P. Diop                    Papakouli Diop   
885  178213.0         E. Capoue                    Etienne Capoue   
886  175254.0        F. Caicedo                    Felipe Caicedo   
887  183899.0         P. Piatti                      Pablo Piatti   
888  173221.0       A. Candreva                  Antonio Candreva   
889  158626.0        M. Debuchy                   Mathieu Debuchy   
890  143001.0          C. Tévez                      Carlos Tévez   
891  173208.0       D. Criscito                 Domenico Criscito   
892  143121.0           J. Sosa                         José Sosa   
893  154472.0      Jorge Molina                Jorge Molina Vidal   
894  146296.0  Andrés Fernández   Andrés Eduardo Fernández Moreno   
895  171791.0        José Fonte        José Miguel da Rocha Fonte   
896  171897.0       A. Guardado                   Andrés Guardado   
897  172962.0          V. Moses                      Victor Moses   
898  142707.0          L. López                    Lisandro López   

                                             PlayerUrl          Nationality  \
793  http://sofifa.com/player/110677/miguel-angel-m...                Spain   
847  http://sofifa.com/player/166706/martin-skrtel/...             Slovakia   
848   http://sofifa.com/player/189068/bas-dost/210005/          Netherlands   
860  http://sofifa.com/player/198031/sergio-miguel-...             Portugal   
861  http://sofifa.com/player/190243/marwin-hitz/21...          Switzerland   
862  http://sofifa.com/player/194222/victor-laguard...                Spain   
863  http://sofifa.com/player/194209/youssef-el-ara...              Morocco   
864  http://sofifa.com/player/193601/german-pezzell...            Argentina   
865  http://sofifa.com/player/193352/ricardo-rodrig...          Switzerland   
866  http://sofifa.com/player/193152/iago-herrerin-...                Spain   
867  http://sofifa.com/player/192546/enda-stevens/2...  Republic of Ireland   
868  http://sofifa.com/player/192123/chris-wood/210...          New Zealand   
869  http://sofifa.com/player/192086/alexandr-kokor...               Russia   
871  http://sofifa.com/player/192045/luis-rodriguez...               Mexico

In [583]:
df.update(feet)

In [584]:
df.Height.unique()

array(['170cm', '187cm', '188cm', '181cm', '175cm', '184cm', '191cm',
       '178cm', '193cm', '185cm', '199cm', '173cm', '168cm', '176cm',
       '177cm', '183cm', '180cm', '189cm', '179cm', '195cm', '172cm',
       '182cm', '186cm', '192cm', '165cm', '194cm', '167cm', '196cm',
       '163cm', '190cm', '174cm', '169cm', '171cm', '197cm', '200cm',
       '166cm', '164cm', '198cm', '155cm', '201cm', '158cm', '162cm',
       '161cm', '160cm', '203cm', '157cm', '156cm', '202cm', '159cm',
       '206cm'], dtype=object)

## Since the Height column has been changed to "cm", we have to work on the weight column

# Weight Cleaning

In [585]:
df.head(2)

ID               Name                      FullName  \
0  158023.0           L. Messi                  Lionel Messi   
1   20801.0  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   

                                           PlayerUrl Nationality   Age  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina  33.0   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal  35.0   

   OverallRating   POT          Club     Contract   Positions Height Weight  \
0           93.0  93.0  FC Barcelona  2004 - 2021  RW| ST| CF  170cm   72kg   
1           92.0  92.0      Juventus  2018 - 2022      ST| LW  187cm   83kg   

  Preferred Foot   BOV Best Position     Joined    Value   Wage  \
0           Left  93.0            RW 2004-07-01  €103.5M  €560K   
1          Right  92.0            ST 2018-07-10     €63M  €220K   

  Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0        €138.4M      429.0      85.0       95.0              70.0   
1         €75.9M      437.0      84.0       95.0              90.0   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0           91.0     88.0  470.0       96.0   93.0         94.0          91.0   
1           82.0     86.0  414.0       88.0   81.0         76.0          77.0   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0          96.0     451.0          91.0          80.0     91.0       94.0   
1          92.0     431.0          87.0          91.0     87.0       95.0   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0     95.0  389.0        86.0     68.0     72.0      69.0        94.0   
1     71.0  444.0        94.0     95.0     84.0      78.0        93.0   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0      347.0        44.0           40.0         93.0    95.0       75.0   
1      353.0        63.0           29.0         95.0    82.0       84.0   

   Composure  Defending  Marking  Standing Tackle  Sliding Tackle  \
0       96.0       91.0     32.0             35.0            24.0   
1       95.0       84.0     28.0             32.0            24.0   

   Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  \
0         54.0        6.0         11.0        15.0            14.0   
1         58.0        7.0         11.0        15.0            14.0   

   GK Reflexes  Total Stats  Base Stats W/F SM     A/W  D/W  IR   PAC   SHO  \
0          8.0       2231.0       466.0  4   4  Medium  Low  5   85.0  92.0   
1         11.0       2221.0       464.0  4   5    High  Low  5   89.0  93.0   

    PAS   DRI   DEF   PHY   Hits  
0  91.0  95.0  38.0  65.0  771.0  
1  81.0  89.0  35.0  77.0  562.0

In [586]:
df.Weight.unique()

array(['72kg', '83kg', '87kg', '70kg', '68kg', '80kg', '71kg', '91kg',
       '73kg', '85kg', '92kg', '69kg', '84kg', '96kg', '81kg', '82kg',
       '75kg', '86kg', '89kg', '74kg', '76kg', '64kg', '78kg', '90kg',
       '66kg', '60kg', '94kg', '79kg', '67kg', '65kg', '59kg', '61kg',
       '93kg', '88kg', '97kg', '77kg', '62kg', '63kg', '95kg', '100kg',
       '58kg', '183lbs', '179lbs', '172lbs', '196lbs', '176lbs', '185lbs',
       '170lbs', '203lbs', '168lbs', '161lbs', '146lbs', '130lbs',
       '190lbs', '174lbs', '148lbs', '165lbs', '159lbs', '192lbs',
       '181lbs', '139lbs', '154lbs', '157lbs', '163lbs', '98kg', '103kg',
       '99kg', '102kg', '56kg', '101kg', '57kg', '55kg', '104kg', '107kg',
       '110kg', '53kg', '50kg', '54kg', '52kg'], dtype=object)

## create a df for the weights measured in lbs

In [587]:
lbs = df[df['Weight'].str.contains('lbs')]

In [588]:
lbs

ID              Name                          FullName  \
793  110677.0              Moyá           Miguel Ángel Moyá Rumbo   
847  166706.0         M. Škrtel                     Martin Škrtel   
848  189068.0           B. Dost                          Bas Dost   
860  198031.0   Sérgio Oliveira  Sérgio Miguel Relvas de Oliveira   
861  190243.0           M. Hitz                       Marwin Hitz   
862  194222.0         Laguardia         Víctor Laguardia Cisneros   
863  194209.0       Y. El Arabi                  Youssef El Arabi   
864  193601.0       G. Pezzella                   Germán Pezzella   
865  193352.0      R. Rodríguez                 Ricardo Rodríguez   
866  193152.0          Herrerín              Iago Herrerín Buisán   
867  192546.0        E. Stevens                      Enda Stevens   
868  192123.0           C. Wood                        Chris Wood   
869  192086.0        A. Kokorin                  Alexandr Kokorin   
871  192045.0      L. Rodríguez                    Luis Rodríguez   
872  189280.0       A. Westwood                   Ashley Westwood   
873  196978.0         C. Wilson                     Callum Wilson   
874  188988.0        M. Lanzini                    Manuel Lanzini   
875  173909.0        K. Boateng              Kevin-Prince Boateng   
876  188427.0          S. Lulić                       Senad Lulić   
877  140293.0        A. Mirante                   Antonio Mirante   
878  188268.0         L. Acosta                    Lautaro Acosta   
879  188042.0         M. Suárez                     Matías Suárez   
880  188041.0    Rafael Carioca           Rafael de Souza Pereira   
881  186547.0      M. Musacchio                   Mateo Musacchio   
883  183353.0       Víctor Díaz          Víctor David Díaz Miguel   
884  178250.0           P. Diop                    Papakouli Diop   
885  178213.0         E. Capoue                    Etienne Capoue   
886  175254.0        F. Caicedo                    Felipe Caicedo   
887  183899.0         P. Piatti                      Pablo Piatti   
888  173221.0       A. Candreva                  Antonio Candreva   
889  158626.0        M. Debuchy                   Mathieu Debuchy   
890  143001.0          C. Tévez                      Carlos Tévez   
891  173208.0       D. Criscito                 Domenico Criscito   
892  143121.0           J. Sosa                         José Sosa   
893  154472.0      Jorge Molina                Jorge Molina Vidal   
894  146296.0  Andrés Fernández   Andrés Eduardo Fernández Moreno   
895  171791.0        José Fonte        José Miguel da Rocha Fonte   
896  171897.0       A. Guardado                   Andrés Guardado   
897  172962.0          V. Moses                      Victor Moses   
898  142707.0          L. López                    Lisandro López   

                                             PlayerUrl          Nationality  \
793  http://sofifa.com/player/110677/miguel-angel-m...                Spain   
847  http://sofifa.com/player/166706/martin-skrtel/...             Slovakia   
848   http://sofifa.com/player/189068/bas-dost/210005/          Netherlands   
860  http://sofifa.com/player/198031/sergio-miguel-...             Portugal   
861  http://sofifa.com/player/190243/marwin-hitz/21...          Switzerland   
862  http://sofifa.com/player/194222/victor-laguard...                Spain   
863  http://sofifa.com/player/194209/youssef-el-ara...              Morocco   
864  http://sofifa.com/player/193601/german-pezzell...            Argentina   
865  http://sofifa.com/player/193352/ricardo-rodrig...          Switzerland   
866  http://sofifa.com/player/193152/iago-herrerin-...                Spain   
867  http://sofifa.com/player/192546/enda-stevens/2...  Republic of Ireland   
868  http://sofifa.com/player/192123/chris-wood/210...          New Zealand   
869  http://sofifa.com/player/192086/alexandr-kokor...               Russia   
871  http://sofifa.com/player/192045/luis-rodriguez...               Mexico

## Remove the lbs

In [589]:
change = lambda x: x.replace("lbs"," ")

In [590]:
lbs.Weight = lbs.Weight.apply(change)

C:\Users\user\AppData\Local\Temp\ipykernel_1608\3961284165.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  lbs.Weight = lbs.Weight.apply(change)


In [591]:
# change the values to float values so we can add it up to kilogram
lbs.Weight = lbs.Weight.astype(float)

C:\Users\user\AppData\Local\Temp\ipykernel_1608\882314962.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  lbs.Weight = lbs.Weight.astype(float)


In [592]:
# add up with the kg value to turn lbs to kilogram
lbs_to_kg = lambda x: x / 2.2046

In [593]:
lbs.Weight = lbs.Weight.apply(lbs_to_kg).astype('int')

C:\Users\user\AppData\Local\Temp\ipykernel_1608\3194326651.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  lbs.Weight = lbs.Weight.apply(lbs_to_kg).astype('int')


In [594]:
#change from 
lbs.Weight = lbs.Weight.astype('str')

C:\Users\user\AppData\Local\Temp\ipykernel_1608\3679387404.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  lbs.Weight = lbs.Weight.astype('str')


In [595]:
new_kg = lambda x: x + 'kg'

In [596]:
lbs.Weight = lbs.Weight.apply(new_kg)

C:\Users\user\AppData\Local\Temp\ipykernel_1608\3271149921.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  lbs.Weight = lbs.Weight.apply(new_kg)


In [597]:
lbs

ID              Name                          FullName  \
793  110677.0              Moyá           Miguel Ángel Moyá Rumbo   
847  166706.0         M. Škrtel                     Martin Škrtel   
848  189068.0           B. Dost                          Bas Dost   
860  198031.0   Sérgio Oliveira  Sérgio Miguel Relvas de Oliveira   
861  190243.0           M. Hitz                       Marwin Hitz   
862  194222.0         Laguardia         Víctor Laguardia Cisneros   
863  194209.0       Y. El Arabi                  Youssef El Arabi   
864  193601.0       G. Pezzella                   Germán Pezzella   
865  193352.0      R. Rodríguez                 Ricardo Rodríguez   
866  193152.0          Herrerín              Iago Herrerín Buisán   
867  192546.0        E. Stevens                      Enda Stevens   
868  192123.0           C. Wood                        Chris Wood   
869  192086.0        A. Kokorin                  Alexandr Kokorin   
871  192045.0      L. Rodríguez                    Luis Rodríguez   
872  189280.0       A. Westwood                   Ashley Westwood   
873  196978.0         C. Wilson                     Callum Wilson   
874  188988.0        M. Lanzini                    Manuel Lanzini   
875  173909.0        K. Boateng              Kevin-Prince Boateng   
876  188427.0          S. Lulić                       Senad Lulić   
877  140293.0        A. Mirante                   Antonio Mirante   
878  188268.0         L. Acosta                    Lautaro Acosta   
879  188042.0         M. Suárez                     Matías Suárez   
880  188041.0    Rafael Carioca           Rafael de Souza Pereira   
881  186547.0      M. Musacchio                   Mateo Musacchio   
883  183353.0       Víctor Díaz          Víctor David Díaz Miguel   
884  178250.0           P. Diop                    Papakouli Diop   
885  178213.0         E. Capoue                    Etienne Capoue   
886  175254.0        F. Caicedo                    Felipe Caicedo   
887  183899.0         P. Piatti                      Pablo Piatti   
888  173221.0       A. Candreva                  Antonio Candreva   
889  158626.0        M. Debuchy                   Mathieu Debuchy   
890  143001.0          C. Tévez                      Carlos Tévez   
891  173208.0       D. Criscito                 Domenico Criscito   
892  143121.0           J. Sosa                         José Sosa   
893  154472.0      Jorge Molina                Jorge Molina Vidal   
894  146296.0  Andrés Fernández   Andrés Eduardo Fernández Moreno   
895  171791.0        José Fonte        José Miguel da Rocha Fonte   
896  171897.0       A. Guardado                   Andrés Guardado   
897  172962.0          V. Moses                      Victor Moses   
898  142707.0          L. López                    Lisandro López   

                                             PlayerUrl          Nationality  \
793  http://sofifa.com/player/110677/miguel-angel-m...                Spain   
847  http://sofifa.com/player/166706/martin-skrtel/...             Slovakia   
848   http://sofifa.com/player/189068/bas-dost/210005/          Netherlands   
860  http://sofifa.com/player/198031/sergio-miguel-...             Portugal   
861  http://sofifa.com/player/190243/marwin-hitz/21...          Switzerland   
862  http://sofifa.com/player/194222/victor-laguard...                Spain   
863  http://sofifa.com/player/194209/youssef-el-ara...              Morocco   
864  http://sofifa.com/player/193601/german-pezzell...            Argentina   
865  http://sofifa.com/player/193352/ricardo-rodrig...          Switzerland   
866  http://sofifa.com/player/193152/iago-herrerin-...                Spain   
867  http://sofifa.com/player/192546/enda-stevens/2...  Republic of Ireland   
868  http://sofifa.com/player/192123/chris-wood/210...          New Zealand   
869  http://sofifa.com/player/192086/alexandr-kokor...               Russia   
871  http://sofifa.com/player/192045/luis-rodriguez...               Mexico

In [598]:
df.update(lbs)

In [599]:
df.Weight.unique()

array(['72kg', '83kg', '87kg', '70kg', '68kg', '80kg', '71kg', '91kg',
       '73kg', '85kg', '92kg', '69kg', '84kg', '96kg', '81kg', '82kg',
       '75kg', '86kg', '89kg', '74kg', '76kg', '64kg', '78kg', '90kg',
       '66kg', '60kg', '94kg', '79kg', '67kg', '65kg', '59kg', '61kg',
       '93kg', '88kg', '97kg', '77kg', '62kg', '63kg', '95kg', '100kg',
       '58kg', '98kg', '103kg', '99kg', '102kg', '56kg', '101kg', '57kg',
       '55kg', '104kg', '107kg', '110kg', '53kg', '50kg', '54kg', '52kg'],
      dtype=object)

In [604]:
df.head()

ID               Name                      FullName  \
0  158023.0           L. Messi                  Lionel Messi   
1   20801.0  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389.0           J. Oblak                     Jan Oblak   
3  192985.0       K. De Bruyne               Kevin De Bruyne   
4  190871.0          Neymar Jr    Neymar da Silva Santos Jr.   

                                           PlayerUrl Nationality  Age  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28   

   OverallRating   POT                 Club     Contract   Positions Height  \
0           93.0  93.0         FC Barcelona  2004 - 2021  RW| ST| CF  170cm   
1           92.0  92.0             Juventus  2018 - 2022      ST| LW  187cm   
2           91.0  93.0      Atlético Madrid  2014 - 2023          GK  188cm   
3           91.0  91.0      Manchester City  2015 - 2023     CAM| CM  181cm   
4           91.0  91.0  Paris Saint-Germain  2017 - 2022     LW| CAM  175cm   

  Weight Preferred Foot   BOV Best Position     Joined    Value   Wage  \
0   72kg           Left  93.0            RW 2004-07-01  €103.5M  €560K   
1   83kg          Right  92.0            ST 2018-07-10     €63M  €220K   
2   87kg          Right  91.0            GK 2014-07-16    €120M  €125K   
3   70kg          Right  91.0           CAM 2015-08-30    €129M  €370K   
4   68kg          Right  91.0            LW 2017-08-03    €132M  €270K   

  Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0        €138.4M      429.0      85.0       95.0              70.0   
1         €75.9M      437.0      84.0       95.0              90.0   
2        €159.4M       95.0      13.0       11.0              15.0   
3          €161M      407.0      94.0       82.0              55.0   
4        €166.5M      408.0      85.0       87.0              62.0   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0           91.0     88.0  470.0       96.0   93.0         94.0          91.0   
1           82.0     86.0  414.0       88.0   81.0         76.0          77.0   
2           43.0     13.0  109.0       12.0   13.0         14.0          40.0   
3           94.0     82.0  441.0       88.0   85.0         83.0          93.0   
4           87.0     87.0  448.0       95.0   88.0         89.0          81.0   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0          96.0     451.0          91.0          80.0     91.0       94.0   
1          92.0     431.0          87.0          91.0     87.0       95.0   
2          30.0     307.0          43.0          60.0     67.0       88.0   
3          92.0     398.0          77.0          76.0     78.0       91.0   
4          95.0     453.0          94.0          89.0     96.0       91.0   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0     95.0  389.0        86.0     68.0     72.0      69.0        94.0   
1     71.0  444.0        94.0     95.0     84.0      78.0        93.0   
2     49.0  268.0        59.0     78.0     41.0      78.0        12.0   
3     76.0  408.0        91.0     63.0     89.0      74.0        91.0   
4     83.0  357.0        80.0     62.0     81.0      50.0        84.0   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0      347.0        44.0           40.0         93.0    95.0       75.0   
1      353.0        63.0           29.0         95.0    82.0       84.0   
2      140.0        34.0           19.0         11.0    65.0       11.0   
3      408.0        76.0           66.0         88.0    94.0       84.0   
4      356.0        51.0           36.0         87.0    90.0       92.0   

   C